test/Concurrency/async_throwing.swift

```diff
@@ -0,0 +1,138 @@
+  let _ = invokeAuto(42) // expected-error{{'async' in a function that does not support concurrency}}
+  let _ = invokeAuto("intuitive") // expected-error{{'async' in a function that does not support concurrency}}
+  
+  let _ = try! asyncRethrows(fn: throwingTask) // expected-error{{'async' in a function that does not support concurrency}}
+  let _ = try? invoke(fn: throwingTask) // expected-error{{'async' in a function that does not support concurrency}}
+  do {
+   let _ = try invoke(fn: throwingTask) // expected-error{{'async' in a function that does not support concurrency}}
+   let _ = try asyncThrows() // expected-error{{'async' in a function that does not support concurrency}}
+  } catch {
+    // ignore it
+  }
+}
+
+
+func asyncTest() async {
+  ///////////
+  // tests that also omit await
+
+  let _ = invoke(fn: normalTask) // expected-error{{call is 'async' but is not marked with 'await'}}
+  let _ = asyncRethrows(fn: normalTask) // expected-error{{call is 'async' but is not marked with 'await'}}
+  let _ = invokeAuto(42) // expected-error{{call is 'async' but is not marked with 'await'}}
+
+  // expected-error@+2 {{call can throw, but it is not marked with 'try' and the error is not handled}}
+  // expected-error@+1 {{call is 'async' but is not marked with 'await'}}
+  let _ = asyncThrows()
+  
+  // expected-note@+3{{call is to 'rethrows' function, but argument function can throw}}
+  // expected-error@+2{{call can throw, but it is not marked with 'try' and the error is not handled}}
+  // expected-error@+1{{call is 'async' but is not marked with 'await'}}
+  let _ = invoke(fn: throwingTask)
+
+  ///////////
+  // tests that use await and handles the exceptions
+
+  // expected-note@+2{{call is to 'rethrows' function, but argument function can throw}}
+  // expected-error@+1{{call can throw, but it is not marked with 'try' and the error is not handled}}
+  let _ = await invoke(fn: throwingTask)
+  let _ = await invoke(fn: normalTask) // ok
+
+  let _ = await asyncRethrows(fn: normalTask)
+  let _ = await try! asyncRethrows(fn: normalTask) // expected-warning{{no calls to throwing functions occur within 'try' expression}}
+  let _ = await try? asyncRethrows(fn: normalTask) // expected-warning{{no calls to throwing functions occur within 'try' expression}}
+  
+  let _ = await try! asyncRethrows(fn: throwingTask)
+  let _ = await try? asyncRethrows(fn: throwingTask)
+  let _ = await try! asyncThrows()
+  let _ = await try? asyncThrows()
+
+  //////////
+  // some auto-closure tests
+
+  let _ = await invokeAuto("intuitive")
+  let _ = await try! invokeAuto(await throwingTask())
+  let _ = await try? invokeAuto(await throwingTask())
+  let _ = await invokeAuto(await try! throwingTask())
+  let _ = await invokeAuto(await try? throwingTask())
+
+  // FIXME: we currently emit a less-helpful error message here.
+  // it would be better if we said "call is 'async' in an autoclosure argument that is not marked with 'await'"
```

Comment:
Should I also try to fix this error message in this PR?

Reply:
It would be nice to get this in since it cleans up many callsites of tasks, and work on follow ups separately perhaps?

Reply:
It ended up being simple enough to add to this PR.